src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -76,20 +77,49 @@ public void WaitForQueue_TestOnly()
             queue.Wait();
         }
 
-        public IContext CreateContext()
-            => new Context(this, null);
+        public IContext CreateContext(params WatchedDirectory[] watchedDirectories)
+        {
+            return new Context(this, watchedDirectories.ToImmutableArray());
+        }
 
         /// <summary>
-        /// Creates an <see cref="IContext"/> that watches all files in a directory, in addition to any files explicitly requested by <see cref="IContext.EnqueueWatchingFile(string)"/>.
+        /// Gives a hint to the <see cref="IContext"/> that we should watch a top-level directory for all changes in addition
+        /// to any files called by <see cref="IContext.EnqueueWatchingFile(string)"/>.
         /// </summary>
-        public IContext CreateContextForDirectory(string directoryFilePath)
+        /// <remarks>
+        /// This is largely intended as an optimization; consumers should still call <see cref="IContext.EnqueueWatchingFile(string)" />
+        /// for files they want to watch. This allows the caller to give a hint that it is expected that most of the files being
+        /// watched is under this directory, and so it's more efficient just to watch _all_ of the changes in that directory
+        /// rather than creating and tracking a bunch of file watcher state for each file separately. A good example would be
+        /// just creating a single directory watch on the root of a project for source file changes: rather than creating a file watcher
+        /// for each individual file, we can just watch the entire directory and that's it.
+        /// </remarks>
+        public sealed class WatchedDirectory
         {
-            if (directoryFilePath == null)
+            public WatchedDirectory(string path, string? extensionFilter)
             {
-                throw new ArgumentNullException(nameof(directoryFilePath));
+                // We are doing string comparisons with this path, so ensure it has a trailing \ so we don't get confused with sibling
+                // paths that won't actually be covered.
+                if (!path.EndsWith("\\"))
+                {
+                    path += "\\";
```

Comment:
1. this is ok because we're windows?
2. should we use FileSystem.Separator (or whatever the bcl exposes)?